playbook/actions/gitops.go

```diff
@@ -0,0 +1,236 @@
+		if err != nil {
+			return nil, err
+		}
+		response.CreatedPR = prNumber
+	}
+
+	return &response, nil
+}
+
+// generateSpec generates the spec for the git client from the action
+func (t *GitOps) generateSpec(ctx context.Context, action v1.GitOpsAction) error {
+	var err error
+
+	if action.Repo.Base == "" {
+		action.Repo.Base = "master"
+	}
+
+	if action.Repo.Branch == "" {
+		action.Repo.Branch = action.Repo.Base
+	} else {
+		action.Repo.Branch, err = gomplate.RunTemplate(t.env.AsMap(), gomplate.Template{Template: action.Repo.Branch})
+		if err != nil {
+			return err
+		}
+	}
+
+	t.spec = &connectors.GitopsAPISpec{
+		Repository: action.Repo.URL,
+		Base:       action.Repo.Base,
+		Branch:     action.Repo.Branch,
+		CommitMsg:  action.Commit.Message,
+		User:       action.Commit.AuthorName,
+		Email:      action.Commit.AuthorEmail,
+	}
+
+	if t.spec.CommitMsg != "" {
+		t.spec.CommitMsg, err = gomplate.RunTemplate(t.env.AsMap(), gomplate.Template{Template: t.spec.CommitMsg})
+		if err != nil {
+			return err
+		}
+	}
+
+	if action.Repo.Connection != "" {
+		conn, err := ctx.HydrateConnectionByURL(action.Repo.Connection)
+		if err != nil {
+			return err
+		}
+
+		switch conn.Type {
+		case models.ConnectionTypeGithub:
+			t.spec.GITHUB_TOKEN = conn.Password
+			t.shouldCreatePR = true
+
+		case models.ConnectionTypeAzureDevops:
+			// TODO: Azure devops connection doesn't have git credentials ...?
+
+		case models.ConnectionTypeGit:
+			// TODO: need to finalize this once git connection is implemented
+
+		default:
+			return fmt.Errorf("unsupported connection type: %s", conn.Type)
+		}
+	}
+
+	if action.PullRequest == nil {
+		logger.Warnf("no pull request spec was provided on gitops action. So no PR will be created")
+		t.shouldCreatePR = false
+	}
+
+	if t.shouldCreatePR && action.Repo.Base == action.Repo.Branch {
+		logger.Warnf("no base branch was provided on gitops action. So no PR will be created")
+		t.shouldCreatePR = false
+	}
+
+	if t.shouldCreatePR {
+		t.spec.PullRequest = &connectors.PullRequestTemplate{
+			Base:   action.Repo.Base,
+			Branch: action.Repo.Branch,
+			Title:  action.PullRequest.Title,
+			Tags:   action.PullRequest.Tags,
+		}
+
+		t.spec.PullRequest.Title, err = gomplate.RunTemplate(t.env.AsMap(), gomplate.Template{Template: action.PullRequest.Title})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *GitOps) cloneRepo(ctx context.Context, action v1.GitOpsAction) (connectors.Connector, *gitv5.Worktree, error) {
+	connector, workTree, err := git.Clone(ctx, t.spec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return connector, workTree, nil
+}
+
+func (t *GitOps) applyPatches(ctx context.Context, action v1.GitOpsAction) error {
+	for _, patch := range action.Patches {
+		path, err := gomplate.RunTemplate(t.env.AsMap(), gomplate.Template{Template: patch.Path})
+		if err != nil {
+			return err
+		}
+
+		fullpath := filepath.Join(t.workTree.Filesystem.Root(), patch.Path)
```

Comment:
Add a log line to add results.stdout: 'Patching {file}' - I think we should also unglob